fix(nuxt): improve default router `scrollBehaviour`

bernhardberger opened the pull request:
handle first page load and page refresh scenarios (restore scroll position without jumps), fixes hash-navigation when changing pages and browser back/forward positions, especiall when page transitions are involved.

### 🔗 Linked issue

fixes #24941 
fixes #22487 
fixes #19664
fixes #25030

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [X] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

handle first page load and page refresh scenarios (restore scroll position without jumps), fixes hash-navigation when changing pages and browser back/forward positions, especially when page transitions are involved.

### 📝 Checklist

- [X] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

danielroe commented on packages/nuxt/src/pages/runtime/router.options.ts:
```diff
@@ -1,53 +1,77 @@
 import { useNuxtApp } from '#app/nuxt'
-import { isChangingPage } from '#app/components/utils'
 import { useRouter } from '#app/composables/router'
+import { START_LOCATION } from 'vue-router'
```
```suggestion
// @ts-ignore virtual file
import { START_LOCATION } from '#build/pages'
```

danielroe commented on packages/nuxt/src/pages/runtime/router.options.ts:
```diff
@@ -1,53 +1,77 @@
-import type { RouteLocationNormalized, RouterScrollBehavior } from '#vue-router'
 import { nextTick } from 'vue'
 import type { RouterConfig } from 'nuxt/schema'
+import { START_LOCATION } from 'vue-router'
```
```suggestion
// @ts-ignore virtual file
import { START_LOCATION } from '#build/pages'
```

bernhardberger commented:
fyi, fixed the typescript errors and force pushed my branch to avoid commit clutter

manniL commented on packages/nuxt/src/pages/runtime/router.options.ts:
```diff
@@ -11,43 +11,70 @@ type ScrollPosition = Awaited<ReturnType<RouterScrollBehavior>>
-        resolve(position)
+      nuxtApp.hooks.hookOnce(hookToWait, () => {
+        return nextTick(() => {
+          // Without this setTimeout, the scroll behaviour is not applied, this is however working reliably
+          // todo: figure out how we can solve this without setTimeout of 0ms
```
Maybe using `nextTick` twice here? 🤔

bernhardberger replied:
I tried that, unfortunately it doesn't help. I even tried chaining multiple nextTicks() here and it didn't help. I think it has to do with what's happening when using setTimeout: https://medium.com/@sohnu/settimeout-with-time-0-what-does-it-really-mean-3b306880a0f6

bernhardberger replied:
Should have added the quote I'm referring to:

> But what we have to see here is, all the 3 setTimeouts are taken away from the actual call executions and when there is nothing in the actual call stack, the setTimeouts started to execute with the time specified.
> 
> So whenever we want to take the call outside of the flow and run it once after the call stack is free, we can use setTimeout with time 0.

Chalks replied:
I ran into an issue that had the same sort of resolution. I don't really understand what it is about `nextTick` that's not working in every case, but I opened a PR that might be relevant to this discussion: https://github.com/nuxt/nuxt/pull/25817

manniL commented on packages/nuxt/src/pages/runtime/router.options.ts:
```diff
@@ -11,43 +11,70 @@ type ScrollPosition = Awaited<ReturnType<RouterScrollBehavior>>
+      // to avoid page jumps on load
+      window.history.scrollRestoration = 'auto'
+      return savedPosition
+    } else {
+      // For other navigations, ensure manual scrollRestoration
+      window.history.scrollRestoration = 'manual'
+    }
```
```suggestion

     // For other navigations, ensure manual scrollRestoration
    window.history.scrollRestoration = 'manual'
```

manniL commented on packages/nuxt/src/pages/runtime/router.options.ts:
```diff
@@ -11,43 +11,70 @@ type ScrollPosition = Awaited<ReturnType<RouterScrollBehavior>>
-    if (to.path === from.path) {
-      if (from.hash && !to.hash) {
-        return { left: 0, top: 0 }
+    const calculatePosition = (behavior: ScrollBehavior): ScrollPosition => {
```
Could we move this out of the function and pass the `to` as well? 🤔

bernhardberger replied:
Also needs `savedPosition`.

I've moved it out of the function now. I agree with you for the sake of code readability.

bernhardberger committed:
fix: improved default router.options scrollBehaviour

handle first page load and page refresh scenarios (restore scroll position without jumps), fixes hash-navigation when changing pages and browser back/forward positions, especiall when page transitions are involved. fixes #24941, #22487, #25030 and #19664

Signed-off-by: Bernhard Berger <bernhard.berger@gmail.com>

bernhardberger committed:
Merge branch 'main' into fix/default-scroll-behaviour

bernhardberger committed:
Merge branch 'main' into fix/default-scroll-behaviour

bernhardberger committed:
Merge branch 'main' into fix/default-scroll-behaviour

danielroe committed:
Merge branch 'main' into fix/default-scroll-behaviour

danielroe committed:
Merge branch 'main' into fix/default-scroll-behaviour

danielroe commented:
Would you be able to update the PR following https://github.com/nuxt/nuxt/pull/25483? 🙏 

And thank you so much for this ❤️

bernhardberger committed:
Merge branch 'main' into fix/default-scroll-behaviour

bernhardberger commented:
> Would you be able to update the PR following #25483? 🙏
> 
> And thank you so much for this ❤️

I'm a bit limited on time this week. Hopefully I can get to it next week.

danielroe commented:
Marking as draft just for triage - feel free to unmark whenever it's ready from your end 🙏

bernhardberger commented:
> > Would you be able to update the PR following #25483? 🙏
> > And thank you so much for this ❤️
> 
> I'm a bit limited on time this week. Hopefully I can get to it next week.

Turns out I've been drowned in work and family stuff ever since then. I haven't abandoned it, it's still on my mind, however I just have zero spare time at the moment.

danielroe commented:
this PR is waiting on being updated following the merge of https://github.com/nuxt/nuxt/pull/25483 - if you have time to investigate or open a PR (whether built on top of this one or not) that would be fantastic 🙏

bernhardberger commented:
I think some of the stuff of this PR has been addressed already.

Also, the mystery of the `setTimeout` has been solved and explained in another PR (should be linked here somewhere 🤔) and could (probably should) be replaced by waiting for a `requestAnimationFrame`. The other stuff - I think - is still valid (especially the hash and history navigation retoration stuff).

I'm happy helping out with rebasing the PR, but I've unfortunately been down different other rabbitholes and still trying to dog my way up from there, so there's no time to tackle all of it rn by myself. Any help would be greatly appreciated!

rylanharper commented:
Hey all! @danielroe @manniL @bernhardberger 

I tested this PR using `requestAnimationFrame` instead of `setTimeout`, and it resolves the scroll "flash" issue that's been a major visual bug on many SSR Nuxt sites (at least in my local setup). This is after I cleared my local browser Application/Network cache data and also testing via incognito browser. Everything else looks good, no visual errors or console warnings on my end.

Has anyone else had a chance to test it? Do you see the navigation working smoothly with this change? Not sure what else needs to be addressed here.. 

Just checking in to move this PR forward—one of my clients has been asking about it. This really is the only prevalent visual bug in Nuxt and it would be phenomenal to see it fixed. Let me know if there’s anything else that needs testing!

rylanharper commented:
Ah okay got it! Well I tested it out with some changes and at least the flash issue seems to be resolved 😅

danielroe commented:
if you’re happy to open another pr i would of course merge with both of you as co-authors.
